packages/product-json-to-csv/src/writer.js

```diff
@@ -0,0 +1,101 @@
+import archiver from 'archiver'
+import EmitOnce from 'single-emit'
+import fs from 'fs'
+import json2csv from 'json2csv'
+import path from 'path'
+import slugify from 'slugify'
+import tmp from 'tmp'
+import mapHeaders from './map-headers'
+
+// Accept a highland stream and write the output to a single file
+export function writeToSingleCsvFile(productStream, output, logger, headers) {
+  const trimmedHeaders = headers.map(header => header.trim())
+  output.write(`${trimmedHeaders.join(',')}\n`) // Write headers first
+  const columnNames = mapHeaders(trimmedHeaders)
+  productStream
+    .each(product => {
+      const csvData = json2csv({
+        data: product,
+        fields: columnNames,
+        hasCSVColumnTitle: false,
+        quotes: '',
+      })
+      output.write(`${csvData}\n`)
+    })
+    .done(() => {
+      if (output !== process.stdout) output.end()
+
+      logger.info('All products have been written to CSV file')
+    })
+}
+
+// Accept a highland stream and write the output to multiple files per
+// product type, then compress all files to a zip file
+export function writeToZipFile(productStream, output, logger) {
+  const tmpDir = tmp.dirSync({ unsafeCleanup: true }).name
+  tmp.setGracefulCleanup()
+  let currentProductType
+  let fileStream
+  let columnNames
+  const columnNamesCache = {}
+  const streamCache = {}
+  productStream
+    .each(product => {
+      let hasCSVColumnTitle = false
+      // Process this block only if item is a masterVariant and was
+      // not the last processed item
+      if (product.productType && product.productType !== currentProductType) {
+        // get product columnNames from cache or masterVariant.
+        // variants may not have all header fields so should not be relied on
+        if (columnNamesCache[product.productType])
+          columnNames = columnNamesCache[product.productType]
+        else {
+          hasCSVColumnTitle = true
+          columnNames = mapHeaders(Object.keys(product))
+          columnNamesCache[product.productType] = columnNames
+        }
+        currentProductType = product.productType
```

Comment:
Is this like grouping them by product type? What if they dont come in sorted?

Reply:
It almost definitely wouldn't come sorted. [L62 to L66](https://github.com/commercetools/nodejs/pull/326/files/3ecd2c2f940eb11a575140e0e2c267aa46757765#diff-7cc529106f4be69ec3e79b24c55b1fa7R62) uses a cache to decide to which file to write to